tests/postgres_tests/test_aggregates.py

```diff
@@ -233,6 +233,44 @@ def test_string_agg_array_agg_ordering_in_subquery(self):
                 ).order_by('char_field').values_list('char_field', 'agg')
                 self.assertEqual(list(values), expected_result)
 
+    def test_string_agg_filter(self):
+        values = AggregateTestModel.objects.aggregate(
+            stringagg=StringAgg(
+                'char_field',
+                delimiter=';',
+                filter=Q(char_field__endswith='3')
+            )
+        )
+        self.assertEqual(values, {'stringagg': 'Foo3'})
+
+    def test_string_agg_filter_on_m2m(self):
+        AggregateTestModel.objects.all().delete()
+        subagg1 = AggregateTestModel.objects.create(char_field="FooBar")
+        subagg2 = AggregateTestModel.objects.create(char_field="BarBaz")
+
+        aggr1 = AggregateTestModel.objects.create()
+        aggr1.m2m.add(subagg1)
+
+        aggr2 = AggregateTestModel.objects.create()
+        aggr2.m2m.add(subagg1, subagg2)
+
+        subquery = (
+            AggregateTestModel.objects.annotate(
+                _stringagg=StringAgg(
+                    'm2m__char_field',
+                    delimiter=';',
+                    filter=Q(m2m__char_field__startswith="Bar")
+
+                )
+            )
+            .exclude(_stringagg="")
+            .values("id")
+        )
+        self.assertEqual(list(subquery), [{'id': aggr2.id}])
+
+        values = AggregateTestModel.objects.filter(id__in=Subquery(subquery))
+        self.assertEqual(list(values), [aggr2])
```

Comment:
There's likely a way write a test for this without involving a many-to-many relationship which just makes it harder to reason about. From my understanding the issue happens when `Subquery` is involved.